src/kv/idbstore.rs

```diff
@@ -145,3 +155,214 @@ impl Store for IdbStore {
+        }
+    }
+
+    async fn get(&self, key: &str) -> Result<Option<Vec<u8>>> {
+        match self.pending.lock().await.get(key) {
+            Some(v) => Ok(Some(v.to_vec())),
+            None => self.rt.get(key).await,
+        }
+    }
+}
+
+#[async_trait(?Send)]
+impl Write for WriteTransaction {
+    fn as_read<'a>(&'a self) -> &'a dyn Read {
+        self
+    }
+
+    async fn put(&self, key: &str, value: &[u8]) -> Result<()> {
+        self.pending
+            .lock()
+            .await
+            .insert(key.to_string(), value.to_vec());
+        Ok(())
+    }
+
+    async fn commit(self: Box<Self>) -> Result<()> {
+        // Define rollback() to succeed if no writes have occurred, even if
+        // the underlying transaction has exited. Users who expose themselves
+        // to this would notice if they performed any reads after exposing
+        // themselves to a situation where the transaction would autocommit.
+        let pending = self.pending.lock().await;
+        if pending.is_empty() {
+            return Ok(());
+        }
+
+        let store = self.rt.tx.object_store(OBJECT_STORE)?;
+        let mut callbacks = Vec::with_capacity(pending.len());
+        let mut puts: Vec<oneshot::Receiver<()>> = Vec::with_capacity(pending.len());
+        for (key, value) in pending.iter() {
+            let request = store.put_with_key(&js_sys::Uint8Array::from(&value[..]), &key.into())?;
+            let (sender, receiver) = oneshot::channel::<()>();
+            let callback = Closure::once(move || {
+                if let Err(_) = sender.send(()) {
+                    warn!("oneshot send failed");
+                }
+            });
+            request.set_onsuccess(Some(callback.as_ref().unchecked_ref()));
+            callbacks.push(callback);
+            puts.push(receiver);
+        }
+        join_all(puts).await;
+
+        let (lock, cv) = &*self.pair;
+        let state = cv
+            .wait_until(lock.lock().await, |state| *state != WriteState::Open)
+            .await;
+        match self.rt.tx.error() {
+            Some(e) => return Err(format!("{:?}", e).into()),
+            _ => (),
+        }
+        if *state != WriteState::Committed {
+            return Err(StoreError::Str("Transaction aborted".into()));
+        }
+        Ok(())
+    }
+
+    async fn rollback(self: Box<Self>) -> Result<()> {
+        // Define rollback() to succeed if no writes have occurred, even if
+        // the underlying transaction has exited.
+        if self.pending.lock().await.is_empty() {
+            return Ok(());
+        }
+
+        let (lock, cv) = &*self.pair;
+        match *lock.lock().await {
+            WriteState::Committed | WriteState::Aborted => return Ok(()),
+            _ => (),
+        }
+
+        self.rt.tx.abort()?;
+        let state = cv
+            .wait_until(lock.lock().await, |state| *state != WriteState::Open)
+            .await;
+        match self.rt.tx.error() {
```

Comment:
Can't these be written using `if let Some(e) = self.rt.tx.error() {`

Reply:
yep

Reply:
Clippy for the win!
https://github.com/rocicorp/repc/commit/b67ed4c80066c8aef66550e38f44df8e8b21e74f#diff-e83d091a2db9fc52230899a70d97e3a4R299-R300

Reply:
Sweet. At least I picked up some best practice along my way!